### PR TITLE
added a test checks thin write in qemu-img commit

### PR DESCRIPTION
**Pull request by hereischen**
id: 1646843
Signed-off-by: Haotong Chen <hachen@redhat.com>

### CODE REVIEW

**Comment by hereischen** [comment]
(1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64.io-github-autotest-qemu.thin_write_in_qemu_img_commit: PASS (7.15 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.thin_write_in_qemu_img_commit: PASS (7.47 s)


 (1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.thin_write: PASS (9.32 s)

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    :param test: Qemu test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def _get_image_object(tag):
+        """Get QemuImg object by tag."""
+        img_param = params.object_params(tag)
+        img = QemuImg(img_param, img_root_dir, tag)
+        return img
```
Comment (lolyu) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: could just return.

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        """Create an external snapshot based on tag."""
+        logging.info("Create external snapshot %s." % tag)
+        qit = QemuImgTest(test, params, env, snapshot)
+        qit.create_snapshot()
+
+    def _qemu_io(img, cmd):
+        """Run qemu-io cmd to a given img."""
+        logging.info("Run qemu-io %s" % img.image_filename)
+        q = QemuIOSystem(test, params, img.image_filename)
+        q.cmd_output(cmd, 120)
```
Comment (lolyu) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: is there a need to log `qemu-io` output?

File: qemu/tests/thin_write_in_qemu_img_commit.py (hunk 1 above)
Comment (hereischen) [reply qemu/tests/thin_write_in_qemu_img_commit.py]: The `QemuImg` has to be returned, otherwise it can not be referenced later.

Comment (lolyu) [reply qemu/tests/thin_write_in_qemu_img_commit.py]: I mean variable `img` is redundant.

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    def _verify_map_output(output):
+        """"Verify qemu map output."""
+        expected = {
+            "length": int(utils_numeric.normalize_data_size(
+                params["write_size"], "B")),
+            "start": 0, "depth": 0, "zero": True, "data": False}
+        if expected not in json.loads(output.stdout_text):
+            test.fail("There should be no data from 0 to %s." %
+                      params["write_size"])
```
Comment (hellenmao) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: It would be better "The map info is inaccurate from 0 to %s of the %s image file. " ?

Comment (hereischen) [reply qemu/tests/thin_write_in_qemu_img_commit.py]: updated thanks

**Comment by PaulYuuu** [comment]
Acked-by: Yihuang Yu <yihyu@redhat.com>

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from virttest.qemu_storage import QemuImg
+
+from qemu.tests.qemu_disk_img import QemuImgTest
+from qemu.tests.qemu_disk_img import generate_base_snapshot_pair
+
+
+def run(test, params, env):
+    """
+    Check thin write in qemu-img commit.
+
```
Comment (pingl2016) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: No doc for test steps

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        q.cmd_output(cmd, 120)
+
+    def _verify_map_output(output):
+        """"Verify qemu map output."""
+        expected = {
+            "length": int(utils_numeric.normalize_data_size(
+                params["write_size"], "B")),
+            "start": 0, "depth": 0, "zero": True, "data": False}
+        if expected not in json.loads(output.stdout_text):
+            test.fail("The map info is not accurate from 0 to %s of"
```
Comment (pingl2016) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: test.fail("Commit failed, data from 0 to %s are not zero")

File: qemu/tests/thin_write_in_qemu_img_commit.py (hunk 4 above)
Comment (hereischen) [reply qemu/tests/thin_write_in_qemu_img_commit.py]: updated, thanks.

**Comment by hereischen** [comment]
(1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.thin_write_in_qemu_img_commit: PASS (9.65 s)

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+from qemu.tests.qemu_disk_img import QemuImgTest
+from qemu.tests.qemu_disk_img import generate_base_snapshot_pair
+
+
+def run(test, params, env):
+    """
+    Check thin write in qemu-img commit.
+
+    1) qemu-io write data into base image
```
Comment (pingl2016) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: Write 0x1 into the entire base image through qemu-io

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    2) create a external snapshot sn of the base
+    3) qemu-io write data into snapshot image
```
Comment (pingl2016) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: Write specific length zeros at the beginning of the snapshot

File: qemu/tests/thin_write_in_qemu_img_commit.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    4) commit sn back to base
+    5) qemu-img map base
```
Comment (pingl2016) [inline qemu/tests/thin_write_in_qemu_img_commit.py]: Make sure the zeros have been committed into the base image

**Pull request merged**